### PR TITLE
Borovlev Petr HW03

### PR DESCRIPTION
**Pull request by petyb**

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -2,37 +2,115 @@ module Main where
...
+  show (Sqrt expr) = "sqrt(" ++ show expr ++ ")"
+  show (Add expr1 expr2) = "(" ++ show expr1 ++ ") + (" ++ show expr2 ++ ")"
+  show (Sub expr1 expr2) = "(" ++ show expr1 ++ ") - (" ++ show expr2 ++ ")"
+  show (Mul expr1 expr2) = "(" ++ show expr1 ++ ") * (" ++ show expr2 ++ ")"
+  show (Div expr1 expr2) = "(" ++ show expr1 ++ ") / (" ++ show expr2 ++ ")"
+  show (Pow expr1 expr2) = "(" ++ show expr1 ++ ") ^ (" ++ show expr2 ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  a == b = show a == show b
```
Comment (kajigor) [inline Main.hs:27]: This isn't a good way to implement equality.

File: Main.hs [hunk 2]
```diff
@@ -2,37 +2,115 @@ module Main where
...
+  show :: Error -> String
+  show ZeroDivision = "Division by zero error"
+  show NegativeNumberSqrt = "Taking square root from a negative number error"
+
 
 instance Eq Error where
-  (==) = undefined
+  ZeroDivision == ZeroDivision = True
+  NegativeNumberSqrt == NegativeNumberSqrt = True
+  _ == _ = False
```
Comment (kajigor) [inline Main.hs:42]: Whenever your implementation of a type class is trivial, it's better to derive it.

File: Main.hs [hunk 3]
```diff
@@ -2,37 +2,115 @@ module Main where
...
+                  (Right a, Right b) -> Right (a - b)
+                  (Left a, _) -> Left a
+                  (_, Left b) -> Left b
+eval (Mul exp1 exp2) = let
+                  a = eval exp1
+                  b = eval exp2
+                in case (a, b) of
+                  (Right a, Right b) -> Right (a * b)
+                  (Left a, _) -> Left a
+                  (_, Left b) -> Left b
```
Comment (kajigor) [inline Main.hs:69]: There is a lot of code repetition which you can get rid of by introducing a helper function.

File: Main.hs [hunk 4]
```diff
@@ -2,37 +2,115 @@ module Main where
...
+  show (Pow expr1 expr2) = "(" ++ show expr1 ++ ") ^ (" ++ show expr2 ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  a == b = show a == show b
+
 
-data Error
+data Error = ZeroDivision
+           | NegativeNumberSqrt
```
Comment (kajigor) [inline Main.hs:31]: It could also be beneficial to report the sub-expression evaluation of which produced the error.

**Comment by kajigor** [comment]
And make sure github action passes next time.

**Comment by kajigor** [comment]
There are still issues in the code.

**Comment by petyb** [comment]
> There are still issues in the code.
I think I fixed everything you mentioned above

**Review by kajigor — CHANGES_REQUESTED** [review]

File: Main.hs [hunk 5]
```diff
@@ -2,37 +2,94 @@ module Main where
...
+  show (Sqrt expr) = "sqrt(" ++ show expr ++ ")"
+  show (Add expr1 expr2) = "(" ++ show expr1 ++ ") + (" ++ show expr2 ++ ")"
+  show (Sub expr1 expr2) = "(" ++ show expr1 ++ ") - (" ++ show expr2 ++ ")"
+  show (Mul expr1 expr2) = "(" ++ show expr1 ++ ") * (" ++ show expr2 ++ ")"
+  show (Div expr1 expr2) = "(" ++ show expr1 ++ ") / (" ++ show expr2 ++ ")"
+  show (Pow expr1 expr2) = "(" ++ show expr1 ++ ") ^ (" ++ show expr2 ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  a == b = eval a == eval b
```
[the remark this answers is not in the document]
Comment (petyb) [reply Main.hs:27]: it was not mentioned in the task how to compare expressions, kinda made sense to compare them by value

[the remark this answers is not in the document]
Comment (petyb) [reply Main.hs]: fixed

File: Main.hs [hunk 6]
```diff
@@ -2,37 +2,94 @@ module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
+import Distribution.Compat.ResponseFile (expandResponse)
+import Text.XHtml (base, abbr, reset, background)
+import System.Win32 (xBUTTON1)
```
[the remark this answers is not in the document]
Comment (petyb) [reply Main.hs]: fixed

File: Main.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kajigor) [reply Main.hs]: > it was not mentioned in the task how to compare expressions, kinda made sense to compare them by value

It's indeed *a* way to compare expressions. The issue is that it's not scalable: you won't be able to implement such a notion of equality when we introduce variables. Even if you've managed to do it, it won't be that useful. Imagine we've implemented an optimization pass which gets rid of multiplication by 1: `1 * (x - y) -> x - y`. Now we want to write some tests that check whether the optimization is performed correctly. With your notion of equality, these two expressions are indistinguishable, so your test would test nothing. 

`Eq` should express syntactic equality in general, unless there's a specific reason to have a different notion of it. Unfortunately, I'm struggling to find examples of this in published libraries.

File: Main.hs [hunk 7]
```diff
@@ -3,31 +3,81 @@ module Main where
...
 instance Show Error where
-  show = undefined
+  show :: Error -> String
+  show (ZeroDivision expr) = "Division by zero error in the expression " ++ show expr
+  show (NegativeNumberSqrt expr) = "Taking square root from a negative number error in the expression" ++ show expr
 
 instance Eq Error where
-  (==) = undefined
+  (ZeroDivision _) == (ZeroDivision _) = True
+  (NegativeNumberSqrt _) == (NegativeNumberSqrt _) = True
```
Comment (kajigor) [inline Main.hs:35]: This is not a great way to compare errors either: there is a difference between `ZeroDivision (Num 0)` and `ZeroDivision (Minus (Num 1) (Num 1))`.

**Pull request closed without merging**